File: nfl_sim/engine/logic.py
"""All game logic lies here, and is orchestrated by the `apply_outcome` function."""

from nfl_sim.engine.state import (
    _CLK,
    _DEF,
    _DIST,
    _DN,
    _OFF,
    _Q,
    _SC,
    _YL,
    Intent,
    Outcome,
    _GameState,
)


def apply_outcome(state: _GameState, intent: Intent, outcome: Outcome) -> _GameState:
    """The reducer/transition function.

    Also handles fixups on the outcome object:
    - Sets outcome.touchdown when yardline_100 reaches the endzone
    - Zeros outcome.yards_gained for ST plays (nflfastR convention)
    """
    # Apply time first
    new_clock = state[_CLK] - outcome.time_elapsed
    new_quarter = state[_Q]
    if new_clock <= 0:
        new_clock = 900  # 15 min quarters
        new_quarter = state[_Q] + 1

    # Handle field goal (special case - points without TD, changes possession)
    if intent == Intent.FIELD_GOAL:
        # outcome.yards_gained communicates kick distance from the model;
        # consume it for the made/miss check, then zero it for the trace.
        fg_made = state[_YL] - outcome.yards_gained <= 0
        outcome.yards_gained = 0
        score = state[_SC]
        if fg_made:
            offense_idx = 0 if state[_OFF] == "HOME" else 1
            score = (score[0] + 3, score[1]) if offense_idx == 0 else (score[0], score[1] + 3)
        # After FG attempt (made or missed), other team gets ball at their 25
        return (new_quarter, new_clock, state[_DEF], state[_OFF], 1, 10, 75, score)

    # Handle punt (intentional possession change)
    if intent == Intent.PUNT:
        # The punt model predicts yards; use outcome.yards_gained as punt distance.
        punt_distance = outcome.yards_gained
        outcome.yards_gained = 0  # zero for trace (nflfastR convention)

        punt_landing = state[_YL] - punt_distance
        if punt_landing <= 0:
            # Into or past endzone - touchback at the 25
            receiving_yardline_100 = 75
        else:
            # Receiving team gets ball at the landing spot (flipped perspective)
            receiving_yardline_100 = 100 - punt_landing
        return (
            new_quarter,
            new_clock,
            state[_DEF],
            state[_OFF],
            1,
            min(10, receiving_yardline_100),
            receiving_yardline_100,
            state[_SC],
        )

    # Apply yards (TD check uses <= 0, so don't clamp the lower bound here)
    # Upper bound: >100 would be safety - clamp to 99 for now (safety handling TBD)
    new_yardline_100 = min(99, state[_YL] - outcome.yards_gained)

    # Handle touchdown (yardline_100 reached endzone)
    if new_yardline_100 <= 0:
        outcome.touchdown = True
        offense_idx = 0 if state[_OFF] == "HOME" else 1
        sc = state[_SC]
        score = (sc[0] + 7, sc[1]) if offense_idx == 0 else (sc[0], sc[1] + 7)
        # Reset after TD - other team gets ball at their 25
        return (new_quarter, new_clock, state[_DEF], state[_OFF], 1, 10, 75, score)

    # Handle turnover (interception/fumble from outcome model)
    if outcome.turnover:
        flipped_yardline_100 = 100 - new_yardline_100
        return (
            new_quarter,
            new_clock,
            state[_DEF],
            state[_OFF],
            1,
            min(10, flipped_yardline_100),
            flipped_yardline_100,
            state[_SC],
        )

    # Handle first down
    if outcome.yards_gained >= state[_DIST]:
        return (
            new_quarter,
            new_clock,
            state[_OFF],
            state[_DEF],
            1,
            min(10, new_yardline_100),
            new_yardline_100,
            state[_SC],
        )

    # Handle turnover on downs
    if state[_DN] == 4:
        flipped_yardline_100 = 100 - new_yardline_100
        return (
            new_quarter,
            new_clock,
            state[_DEF],
            state[_OFF],
            1,
            min(10, flipped_yardline_100),
            flipped_yardline_100,
            state[_SC],
        )

    # Normal play - advance down
    return (
        new_quarter,
        new_clock,
        state[_OFF],
        state[_DEF],
        state[_DN] + 1,
        state[_DIST] - outcome.yards_gained,
        new_yardline_100,
        state[_SC],
    )


def is_terminal(state: _GameState) -> bool:
    """Game ends after 4 quarters."""
    return state[_Q] > 4

File: sim_rs/src/logic.rs
//! Port of nfl_sim/engine/logic.py — pure game rules.

use crate::state::{GameState, Intent, Outcome, Team};

#[inline]
pub fn is_terminal(s: &GameState) -> bool {
    s.quarter > 4
}

#[inline]
pub fn apply_time(s: GameState, seconds: i16) -> GameState {
    let mut new_clock = s.clock - seconds;
    let mut new_quarter = s.quarter;
    if new_clock <= 0 {
        new_clock = 900;
        new_quarter = s.quarter + 1;
    }
    GameState {
        quarter: new_quarter,
        clock: new_clock,
        ..s
    }
}

/// Mirrors apply_outcome() in logic.py. Mutates `outcome` for touchdown / ST
/// yards zeroing just like the Python reducer does.
pub fn apply_outcome(s: GameState, intent: Intent, outcome: &mut Outcome) -> GameState {
    // Field goal
    if matches!(intent, Intent::FieldGoal) {
        let fg_made = (s.yardline_100 as i16) - outcome.yards_gained <= 0;
        outcome.yards_gained = 0;
        let mut home = s.home_score;
        let mut away = s.away_score;
        if fg_made {
            match s.offense {
                Team::Home => home += 3,
                Team::Away => away += 3,
            }
        }
        return GameState {
            offense: s.offense.flip(),
            down: 1,
            distance: 10,
            yardline_100: 75,
            home_score: home,
            away_score: away,
            ..s
        };
    }

    // Punt
    if matches!(intent, Intent::Punt) {
        let punt_distance = outcome.yards_gained;
        outcome.yards_gained = 0;
        let landing = (s.yardline_100 as i16) - punt_distance;
        let receiving_yl100: u8 = if landing <= 0 {
            75
        } else {
            (100 - landing) as u8
        };
        return GameState {
            offense: s.offense.flip(),
            down: 1,
            distance: 10u8.min(receiving_yl100),
            yardline_100: receiving_yl100,
            ..s
        };
    }

    // Yards applied. Clamp upper bound at 99 (safety TBD).
    let raw_yl = (s.yardline_100 as i16) - outcome.yards_gained;
    let new_yl100_i = raw_yl.min(99);

    // Touchdown
    if new_yl100_i <= 0 {
        outcome.touchdown = true;
        let mut home = s.home_score;
        let mut away = s.away_score;
        match s.offense {
            Team::Home => home += 7,
            Team::Away => away += 7,
        }
        return GameState {
            offense: s.offense.flip(),
            down: 1,
            distance: 10,
            yardline_100: 75,
            home_score: home,
            away_score: away,
            ..s
        };
    }

    let new_yl100 = new_yl100_i as u8;

    // Turnover
    if outcome.is_turnover() {
        let flipped = 100 - new_yl100;
        return GameState {
            offense: s.offense.flip(),
            down: 1,
            distance: 10u8.min(flipped),
            yardline_100: flipped,
            ..s
        };
    }

    // First down
    if outcome.yards_gained >= s.distance as i16 {
        return GameState {
            down: 1,
            distance: 10u8.min(new_yl100),
            yardline_100: new_yl100,
            ..s
        };
    }

    // Turnover on downs
    if s.down == 4 {
        let flipped = 100 - new_yl100;
        return GameState {
            offense: s.offense.flip(),
            down: 1,
            distance: 10u8.min(flipped),
            yardline_100: flipped,
            ..s
        };
    }

    // Normal — advance down
    GameState {
        down: s.down + 1,
        distance: (s.distance as i16 - outcome.yards_gained).max(0) as u8,
        yardline_100: new_yl100,
        ..s
    }
}
